### PR TITLE
fix Issue 17495 - __traits(getParameterStorageClasses) doesn't work w…

### PR DESCRIPTION
**Pull request by WalterBright**
…ith tuples as parameters

blocking https://github.com/dlang/phobos/pull/5474

### CODE REVIEW

File: src/ddmd/traits.d [hunk 1]
```diff
@@ -1021,7 +1021,7 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
         }
         ex = ex.ctfeInterpret();
         auto ii = ex.toUInteger();
-        if (ii >= fparams.dim)
+        if (ii >= Parameter.dim(fparams))
```
Comment (rainers) [inline src/ddmd/traits.d]: I suspect the error message should show the new upper limit, too.

Comment (WalterBright) [inline src/ddmd/traits.d]: good catch

**Comment by andralex** [comment]
So these are the other issue, right?

```d
../../.dub/packages/libasync-0.8.3/libasync/source/libasync/events.d(57,2): Error: destructor 'libasync.events.EventLoop.~this' is not nothrow
../../.dub/packages/libasync-0.8.3/libasync/source/libasync/events.d(57,2): Error: nothrow constructor 'libasync.events.EventLoop.this' may throw
```

**Comment by CyberShadow** [comment]
> So these are the other issue, right?

The fix for the libasync error is here:

https://github.com/dlang/dmd/pull/6896

**Comment by WalterBright** [comment]
> right?

It's the issue for getParameterStorageClassTuples.

It's a bug, regardless, and the fix for it.

**Review by rainers — APPROVED** [review]

**Pull request merged**